src/os/bluestore/BlueStore.cc

```diff
@@ -7822,6 +7817,143 @@ void BlueStore::_fsck_check_pool_statfs(
   }
 }
 
+void BlueStore::_fsck_repair_shared_blobs(
+  BlueStoreRepairer& repairer,
+  shared_blob_2hash_tracker_t& sb_ref_counts,
+  sb_info_space_efficient_map_t& sb_info)
+{
+  auto sb_ref_mismatches = sb_ref_counts.count_non_zero();
+  dout(1) << __func__ << " repairing shared_blobs, ref mismatch estimate: "
+	  << sb_ref_mismatches << dendl;
+  if (!sb_ref_mismatches) // not expected to succeed, just in case
+    return;
+
+  mempool::bluestore_fsck::map<uint64_t, bluestore_extent_ref_map_t> ref_maps;
+  auto it = db->get_iterator(PREFIX_OBJ, KeyValueDB::ITERATOR_NOCACHE);
+  if (it) {
+    CollectionRef c;
+    spg_t pgid;
+    for (it->lower_bound(string()); it->valid(); it->next()) {
+      dout(30) << __func__ << " key "
+	<< pretty_binary_string(it->key()) << dendl;
+      if (is_extent_shard_key(it->key())) {
+	continue;
+      }
+
+      ghobject_t oid;
+      int r = get_key_object(it->key(), &oid);
+      if (r < 0) {
+	continue;
+      }
+
+      if (!c ||
+	oid.shard_id != pgid.shard ||
+	oid.hobj.get_logical_pool() != (int64_t)pgid.pool() ||
+	!c->contains(oid)) {
+	c = nullptr;
+	for (auto& p : coll_map) {
+	  if (p.second->contains(oid)) {
+	    c = p.second;
+	    break;
+	  }
+	}
+	if (!c) {
+	  continue;
+	}
+      }
+      dout(20) << __func__ << " inspecting shared blob refs for col:" << c->cid
+	<< " obj:" << oid << dendl;
+
+      OnodeRef o;
+      o.reset(Onode::decode(c, oid, it->key(), it->value()));
+      o->extent_map.fault_range(db, 0, OBJECT_MAX_SIZE);
+
+      _dump_onode<30>(cct, *o);
+
+      mempool::bluestore_fsck::set<BlobRef> passed_sbs;
+      for (auto& e : o->extent_map.extent_map) {
+	auto& b = e.blob->get_blob();
+	if (b.is_shared() && passed_sbs.count(e.blob) == 0) {
+	  auto sbid = e.blob->shared_blob->get_sbid();
+	  passed_sbs.emplace(e.blob);
+	  bluestore_extent_ref_map_t ref_map_candidate;
+	  bool to_be_updated = false;
+	  auto it = ref_maps.lower_bound(sbid);
+	  bool add_existing = it != ref_maps.end() && it->first == sbid;
+	  for (auto& p : b.get_extents()) {
+	    if (p.is_valid()) {
+	      if (add_existing) {
+		it->second.get(p.offset, p.length);
+	      } else {
+		ref_map_candidate.get(p.offset, p.length);
+		if (!to_be_updated) {
+		  to_be_updated =
+		    !sb_ref_counts.test_all_zero_range(sbid, p.offset, p.length);
+		}
+	      }
+	    }
+	  }
+	  if (to_be_updated) {
```

Comment:
Scenario to consider.
Imagine 2 objects O1 and O2 referencing same blob B, but different sections of it.
When processing O1 we check sb_refs_counts and everything is OK => we do not mark anything in ref_maps.
Then we process O2 and we detect problems with sb_ref_counts => we create ref_maps.

Now we are ready to fix blob B, but our ref_map only has references from O2.

Is that so?

Reply:
Thanks for the good catch! Unfortunately I see no way other than having two passes over all objects.. At the first pass we collect all the potentially broken sbids and on the second - building proper refs_map.